### ISSUE TITLE
[aws-cpp-sdk-s3]: TransferManager threads cause failed assertions at program exit

### ISSUE THREAD

**Issue by grrtrr** [aws/aws-sdk-cpp#2274]
### Abstract

The use of detached asynchronous threads running in the `TransferManager`  interferes with the `Aws::ShutdownAPI` shutdown sequence in multiple cases.

One _existing case_ is #2243, whose partial fix #2244 still has not fully addressed the problem. This issue presents a _second case of failed assertions_ that result from _allowing threads to run_ after `main()` has returned and `Aws::ShutdownAPI` has started or completed.

### Failed assertion at program exit

We are repeatedly seeing failed assertions of this kind:
```console
Fatal error condition occurred in external/aws-c-common/source/allocator.c:209: allocator != ((void *)0)
Exiting Application

#19      at 0x5643bc8cb05a in aws_mem_release
#18      at 0x5643bc8d844b in cJSON_Delete
#17      at 0x5643bc8d8485 in cJSON_Delete
#16      at 0x5643bc8c781a in s_endpoints_ruleset_destroy
#15      at 0x5643bc8ce995 in aws_ref_count_release
#14      at 0x5643bc8c9a42 in aws_endpoints_ruleset_release
#13      at 0x5643bc8c084a in s_endpoints_rule_engine_destroy
#12      at 0x5643bc8ce995 in aws_ref_count_release
#11      at 0x5643bc8c1a12 in aws_endpoints_rule_engine_release
#10      at 0x5643bc11cf46 in std::_Sp_counted_base<>::_M_release()
#9       at 0x5643bc278077 in Aws::S3::S3Client::~S3Client()
#8       at 0x5643bc11cf46 in std::_Sp_counted_base<>::_M_release()
#7       at 0x5643bc25e648 in Aws::Transfer::TransferManager::~TransferManager()
#6       at 0x5643bc11cf46 in std::_Sp_counted_base<>::_M_release()
#5       at 0x5643bc258dc3 in std::_Function_base::_Base_manager<>::_M_manager()
#4       at 0x5643bc244b26 in Aws::S3::Model::GetObjectRequest::~GetObjectRequest()
#3       at 0x5643bc2ef17d in std::_Function_base::_Base_manager<>::_M_manager()
#2       at 0x5643bc8123cd in std::thread::_State_impl<>::~_State_impl()
#1       at 0x7fa5e10c76e8 in <?>
```
These occur at random times. 

### Analysis

The failed assertion is due to a `NULL`  allocator in `aws-c-common`:
```c
// aws-c-common/source/allocator.c
void aws_mem_release(struct aws_allocator *allocator, void *ptr) {
    AWS_FATAL_PRECONDITION(allocator != NULL);  // <== LINE 209
    AWS_FATAL_PRECONDITION(allocator->mem_release != NULL);

    if (ptr != NULL) {
        allocator->mem_release(allocator, ptr);
    }
}
```
This appears at the top of the stack trace, invoked via `cJSON_Delete` in `aws-c-common/source/external/cJSON.c`. This function recursively calls itself before invoking `global_hooks.deallocate`. The hooks are set via `aws_json_module_init`:
```c
// aws-c-common/source/json.c
void aws_json_module_init(struct aws_allocator *allocator) {
    if (!s_aws_json_module_initialized) {
        s_aws_json_module_allocator = allocator;
        struct cJSON_Hooks allocation_hooks = {
            .malloc_fn = s_aws_cJSON_alloc,
            .free_fn = s_aws_cJSON_free,
        };
        cJSON_InitHooks(&allocation_hooks);
        s_aws_json_module_initialized = true;
    }
}
```
The  "deallocate" `.free_fn` hook is `s_aws_cJSON_free`, which wraps `aws_mem_release` at the top of the trace:
```c
// aws-c-common/source/json.c
static void s_aws_cJSON_free(void *ptr) {
    aws_mem_release(s_aws_json_module_allocator, ptr);
}  
```
The _root cause_ of above failed assertion in line 209 is that `s_aws_json_module_allocator` was `NULL` at the time of invocation. There is only one place where it is set to `NULL`, in `aws_json_module_cleanup`:
```c
void aws_json_module_cleanup(void) {
    if (s_aws_json_module_initialized) {
        s_aws_json_module_allocator = NULL;  // <== HERE
        s_aws_json_module_initialized = false;
    }
}
```

`aws_json_module_cleanup` is invoked from `aws_common_library_clean_up`, which is called via `aws-c-sdkutils` `aws_sdkutils_library_clean_up`, which is called from `aws-crt-cpp ApiHandle::~ApiHandle()`, the `ApiHandle` destructor that is invoked when `aws-cpp-sdk-core` deletes its `g_apiHandle` global variable in `Aws::CleanupCrt()`.

#### Timeline of the failed assertion

The `Aws::Transfer::TransferManager::~TransferManager()` `shared_ptr` is kept alive until the last thread associated with any of its `TransferHandles` has exited.

1. At the bottom of the trace is `Aws::S3::Model::GetObjectRequest::~GetObjectRequest()`. The `S3Client::GetRequest` is _synchronous_, the const-ref to its `GetObjectRequest` argument stays alive until the function (request) completes. The `GetObjectAsync` used for the ranged-GET requests is analogous, as it just runs `GetObject` in a (detached) thread of the `S3Client` executor. Since the thread running `GetObject` has a `shared_from_this()` reference to the `TransferManager`, the `TransferManager` will stay alive until its last (asynchronous) transfer has completed.
3. The `Aws::Transfer::TransferManager::~TransferManager()` destructor runs. The `TransferManager` keeps a `shared_ptr` reference to an `Aws::S3::S3Client`.
4. The `Aws::S3::S3Client::S3Client()` is destructed next, which starts the `aws-c-sdkutils` release process via `aws_endpoints_rule_engine_release`.
5. Since this whole sequence happens at some random point of time after `main()` has returned and thus `Aws::CleanupCrt()` either is running or has already completed, `aws_json_module_cleanup` has already been called when `s_aws_cJSON_free` wants to `aws_mem_release` from a now `NULL` `s_aws_json_module_allocator`, causing the failed assertion in line 209 of `aws-c-common/source/allocator.c`.

### Root cause of these problems

Neither this issue nor #2243 would occur if the `S3Client` (along with its `TransferManager`) were discontinued.

Until that is possible, there is no easy solution around the problem, since
- `main()` returns once its control flow has reached the end,
- at that time there may still be asynchronous `TransferManager` requests running to completion (these could have been canceled from the `main` thread, or just ignored);
- both the `PooledThreadExecutor` and the `DefaultExecutor` let the thread-task functions (transfers) run to completion before they allow a cancellation/shutdown;
- due to the `shared_ptr` references kept by the asynchronous-transfer threads, the `TransferManager` _stays alive until the very last transfer (successful or not) has completed_.

As a consequence, until there is a way to block `Aws::ShutdownAPI()` until the last `TransferManager` has exited, or until the `S3Client TransferManager` has been discontinued, _API shutdown needs to be robust against threads that are completing while/after  `Aws::ShutdownAPI()` has run_.

### AWS CPP SDK version used

1.10.18 (problem also present on `master`).

### Compiler and Version used

gcc / clang

### Operating System and version

Linux, ubuntu 18.04

**Comment by jmklix** [comment]
Thanks for bringing this up. The team is reviewing the PR's

**Comment by grrtrr** [comment]
As far as the `TransferManager` is concerned, this issue is fixed by #2291.
